apiserver/common/credentialcommon/modelcredential.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package credentialcommon
+
+import (
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/cloud"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/context"
+)
+
+// ValidateModelCredential checks if a cloud credential is valid for a model.
+func ValidateModelCredential(persisted PersistedCloudEntitiesBackend, provider CloudProvider, callCtx context.ProviderCallContext) (params.ErrorResults, error) {
+	// We only check persisted machines vs known cloud instances.
+	// In the future, this check may be extended to other cloud resources,
+	// entities and operation-level authorisations such as interfaces,
+	// ability to CRUD storage, etc.
+	return CheckMachineInstances(persisted, provider, callCtx)
+}
+
+// ValidateNewModelCredential checks if a new cloud credential can be valid
+// for a given model.
+// Note that this call does not validate credential against the cloud of the model.
+func ValidateNewModelCredential(backend PersistedModelBackend, newEnv NewEnvironFunc, callCtx context.ProviderCallContext, credential *cloud.Credential) (params.ErrorResults, error) {
+	model, err := backend.Model()
+	if err != nil {
+		return failCredentialValidation(errors.Trace(err))
```

Comment:
Can you rename `failCredentialValidation` to `fail` like you used in migrationtarget below? I think the current name is long enough that it obscures what's going on. It needs to be more convenient/readable than the alternative of `return empty, errors.Trace(err)`, otherwise it doesn't pay the cost of having to look the function up to find out what it does.

Reply:
Since this is package level variable, simple "fail" is not descriptive enough. However, I will move this within each func that needs it. This way, it'll be just fail()